combinetf2/tensorwriter.py

```diff
@@ -0,0 +1,814 @@
+            self.dict_logkhalfdiff_indices[channel][name] = {}
+
+        norm = self.get_flat_values(h)
+        sumw2 = self.get_flat_variances(h)
+
+        if not self.allow_negative_expectation:
+            norm = np.maximum(norm, 0.0)
+        if not np.all(np.isfinite(sumw2)):
+            raise RuntimeError(
+                f"{len(sumw2)-sum(np.isfinite(sumw2))} NaN or Inf values encountered in variances for {name}!"
+            )
+        if not np.all(np.isfinite(norm)):
+            raise RuntimeError(
+                f"{len(norm)-sum(np.isfinite(norm))} NaN or Inf values encountered in nominal histogram for {name}!"
+            )
+
+        self.dict_norm[channel][name] = norm
+        self.dict_sumw2[channel] += sumw2
+
+    def add_channel(self, axes, name=None):
+        if name is None:
+            name = f"ch{len(self.channels)}"
+        print(f"Add new channel {name}")
+        self.channels[name] = {"axes": axes}
+        ibins = np.prod([len(a) for a in axes])
+        self.nbinschan[name] = ibins
+        self.dict_norm[name] = {}
+        self.dict_sumw2[name] = np.zeros(ibins)
+        self.dict_logkavg[name] = {}
+        self.dict_logkhalfdiff[name] = {}
+        if self.sparse:
+            self.dict_logkavg_indices[name] = {}
+            self.dict_logkhalfdiff_indices[name] = {}
+
+    def _check_hist_and_channel(self, h, channel, add=True):
+        axes = [a for a in h.axes]
+        if channel not in self.channels.keys():
+            if add:
+                self.add_channel(axes, channel)
+            else:
+                raise RuntimeError(f"Channel {channel} not known!")
+        elif axes != self.channels[channel]["axes"]:
+            raise RuntimeError(
+                f"""
+                Histogram axes different from channel axes of channel {channel}
+                \nHistogram axes: {axes}
+                \nChannel axes: {self.channels[channel]["axes"]}
+                """
+            )
+
+    def add_lnN_systematic(
+        self,
+        name,
+        process,
+        channel,
+        uncertainty,
+        profile=True,
+        groups=None,
+    ):
+        if not isinstance(process, (list, tuple, np.ndarray)):
+            process = [process]
+
+        if not isinstance(uncertainty, (list, tuple, np.ndarray)):
+            uncertainty = [uncertainty]
+
+        if len(uncertainty) != 1 and len(process) != len(uncertainty):
+            raise RuntimeError(
+                f"uncertainty must be either a scalar or list with the same length as the list of processes but len(process) = {len(process)} and len(uncertainty) = {len(uncertainty)}"
+            )
+
+        for p, u in zip(process, uncertainty):
+            norm = self.dict_norm[channel][p]
+            syst = norm * u
+            logk = self.get_logk(syst, norm)
+            self.book_logk_avg(logk, channel, p, name)
+
+        self.book_systematic(name, groups=groups, profile=profile)
+
+    def add_systematic(
```

Comment:
Strongly motivated to allow passing multiple systematics in one histogram (e.g. by specifying syst_axes) but this can be added later.

Reply:
There is already an open issue on this: https://github.com/WMass/combinetf2/issues/5